R/g.impute.R

```diff
@@ -95,8 +96,72 @@ g.impute = function(M, I, params_cleaning = c(), desiredtz = "",
   lastmidnight = dmidn$lastmidnight;    lastmidnighti = dmidn$lastmidnighti
   midnights = dmidn$midnights;          midnightsi = dmidn$midnightsi
   #===================================================================
-  # Select data based on strategy
-  if (params_cleaning[["strategy"]] == 1) { 	#protocol based data selection
+  # Trim data based on study_dates_file
+  study_dates_log_used = FALSE
+  if (!is.null(params_cleaning[["study_dates_file"]])) {
+    # Read content of study dates file 
+    studyDates = data.table::fread(file = params_cleaning[["study_dates_file"]], data.table = FALSE)
+    # Check ID and date formats
+    studyDates = check_log(log = studyDates, 
+                           dateformat = params_cleaning[["study_dates_dateformat"]], 
+                           colid = 1, datecols = 2:3, 
+                           logPath = params_cleaning[["study_dates_file"]],
+                           logtype = "study dates log")
+    # identify first and last study dates
+    rowID = which(studyDates[, 1] == ID)
+    if (length(rowID) > 0) {
+      if (length(rowID) > 1) {
+        # this is in the unlikely event that an ID appears twice in the log
+        rowID = rowID[1]
+        warning(paste0("The ID ", ID, " appears twice in the study dates log"), call. = FALSE)
+      }
+      # expected first and last midnight
+      firstmidnight = as.Date(studyDates[rowID, 2], format = params_cleaning[["study_dates_dateformat"]])
+      lastmidnight = as.Date(studyDates[rowID, 3], format = params_cleaning[["study_dates_dateformat"]])
+      # find the dates in metalong
+      firstmidnighti = midnightsi[grep(firstmidnight, midnights)]
+      lastmidnighti = midnightsi[grep(lastmidnight, midnights) + 1] # plus 1 to include the reported date in log
+      # trim data at the beginning
+      if (length(firstmidnighti) > 0) {
+        r4[1:(firstmidnighti - 1)] = 1
+        study_dates_log_used = TRUE
+      } else {
+        # if midnight timestamp for the date is not available, 
+        # do not trim the data and recover the firstmidnight value
+        firstmidnighti = dmidn$firstmidnighti
+        firstmidnight = dmidn$firstmidnight
+        warning(paste0("The start date provided in the study dates file for ID = ", 
```

Comment:
When the start and end date represent the start and end date of the recording this warning is shown even though the part2_daysummary shows that those dates are in the recording. I think the confusion is that you use midnight dates to test whether a date is in the recording. It may be easier to not show this warning. It is up to the user to check whether dates are correct and fall within the recording. For sleeplog we also do not provide warnings when dates are not in the recording.

Reply:
For now I commented the warning out, just in case in the future we want to provide warnings in logs when dates are not coherent with the data recorded.